pennylane/gradients/parameter_shift.py

```diff
@@ -0,0 +1,450 @@
+        function to be applied to the evaluated tapes.
+
+    For a variational evolution :math:`U(\mathbf{p}) \vert 0\rangle` with
+    :math:`N` parameters :math:`\mathbf{p}`,
+    consider the expectation value of an observable :math:`O`:
+
+    .. math::
+
+        f(\mathbf{p})  = \langle \hat{O} \rangle(\mathbf{p}) = \langle 0 \vert
+        U(\mathbf{p})^\dagger \hat{O} U(\mathbf{p}) \vert 0\rangle.
+
+
+    The gradient of this expectation value can be calculated using :math:`2N` expectation
+    values using the parameter-shift rule:
+
+    .. math::
+
+        \frac{\partial f}{\partial \mathbf{p}} = \frac{1}{2\sin s} \left[ f(\mathbf{p} + s) -
+        f(\mathbf{p} -s) \right].
+
+    **Gradients of variances**
+
+    For a variational evolution :math:`U(\mathbf{p}) \vert 0\rangle` with
+    :math:`N` parameters :math:`\mathbf{p}`,
+    consider the variance of an observable :math:`O`:
+
+    .. math::
+
+        g(\mathbf{p})=\langle \hat{O}^2 \rangle (\mathbf{p}) - [\langle \hat{O}
+        \rangle(\mathbf{p})]^2.
+
+    We can relate this directly to the parameter-shift rule by noting that
+
+    .. math::
+
+        \frac{\partial g}{\partial \mathbf{p}}= \frac{\partial}{\partial
+        \mathbf{p}} \langle \hat{O}^2 \rangle (\mathbf{p})
+        - 2 f(\mathbf{p}) \frac{\partial f}{\partial \mathbf{p}}.
+
+    This results in :math:`4N + 1` evaluations.
+
+    In the case where :math:`O` is involutory (:math:`\hat{O}^2 = I`), the first term in the above
+    expression vanishes, and we are simply left with
+
+    .. math::
+
+      \frac{\partial g}{\partial \mathbf{p}} = - 2 f(\mathbf{p})
+      \frac{\partial f}{\partial \mathbf{p}},
+
+    allowing us to compute the gradient using :math:`2N + 1` evaluations.
+
+    **Example**
+
+    >>> with qml.tape.QuantumTape() as tape:
+    ...     qml.RX(params[0], wires=0)
+    ...     qml.RY(params[1], wires=0)
+    ...     qml.RX(params[2], wires=0)
+    ...     qml.expval(qml.PauliZ(0))
+    ...     qml.var(qml.PauliZ(0))
+    >>> tape.trainable_params = {0, 1, 2}
+    >>> gradient_tapes, fn = qml.gradients.param_shift(tape)
+    >>> res = dev.batch_execute(gradient_tapes)
+    >>> fn(res)
+    [[-0.38751721 -0.18884787 -0.38355704]
+     [ 0.69916862  0.34072424  0.69202359]]
+    """
+
+    # perform gradient method validation
+    if any(m.return_type is qml.operation.State for m in tape.measurements):
+        raise ValueError("Does not support circuits that return the state")
+
+    _gradient_analysis(tape)
+    gradient_tapes = []
+
+    # TODO: replace the JacobianTape._grad_method_validation
+    # functionality before deprecation.
+    method = "analytic" if fallback_fn is None else "best"
+    diff_methods = tape._grad_method_validation(method)
+
+    if not tape.trainable_params or all(g == "0" for g in diff_methods):
+        # Either all parameters have grad method 0, or there are no trainable
+        # parameters.

```

Comment:
Maybe we could save a comment here?
```suggestion
    all_params_grad_method_zero = all(g == "0" for g in diff_methods)
    if not tape.trainable_params or all_params_grad_method_zero:
```